crates/ir/src/insn.rs

```diff
@@ -376,6 +418,114 @@ impl InsnData {
     }
 }
 
+pub struct DisplayInsnData<'a> {
+    insn: Insn,
+    dfg: &'a DataFlowGraph,
+}
+
+impl<'a> DisplayInsnData<'a> {
+    pub fn new(insn: Insn, dfg: &'a DataFlowGraph) -> Self {
+        Self { insn, dfg }
+    }
+}
+
+impl<'a> fmt::Display for DisplayInsnData<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use InsnData::*;
+        let Self { insn, dfg } = *self;
+        let insn_data = dfg.insn_data(insn);
+        match insn_data {
+            Unary { code, args } => {
+                let v = DisplayArgValues::new(args, dfg);
+                write!(f, "{} v{v}.;", code.as_str())
+            }
+            Binary { code, args } => {
+                let v = DisplayArgValues::new(args, dfg);
+                write!(f, "{} {v};", code.as_str(),)
+            }
+            Cast { code, args, ty } => {
+                let v = DisplayArgValues::new(args, dfg);
+                let display_ty = DisplayType::new(*ty, dfg);
+                write!(f, "{} {v} {display_ty};", code.as_str())
+            }
+            Load { args, loc } => {
+                let v = DisplayArgValues::new(args, dfg);
+                write!(f, "load {v} {loc};")
+            }
+            Store { args, loc } => {
+                let v = DisplayArgValues::new(args, dfg);
+                write!(f, "store {v} {loc};")
+            }
+            Call {
+                args,
+                func,
+                ret_ty: ty,
+            } => {
+                let v = DisplayArgValues::new(args, dfg);
+                let display_ty = DisplayType::new(*ty, dfg);
+                write!(f, "call %{func} {v} {display_ty};")
+            }
+            Jump { code, dests } => {
+                let block = dests[0];
+                write!(f, "{code} {block};")
+            }
+            Branch { args, dests } => {
+                let v = DisplayArgValues::new(args, dfg);
+                write!(f, "branch {v} {} {};", dests[0], dests[1])
+            }
+            BrTable {
+                args,
+                default,
+                table,
+            } => {
+                let v = DisplayArgValues::new(args, dfg);
+
+                let mut default_string = String::new();
+                if let Some(block) = default {
+                    write!(&mut default_string, "{block} ")?;
+                }
+
+                let mut blocks_string = String::new();
+                write!(&mut blocks_string, "{}", table[0])?;
+                for block in &table[1..] {
+                    write!(&mut blocks_string, ",{block}")?;
+                }
+
+                write!(f, "br_table {v} {default_string} {blocks_string};")
```

Comment:
I guess we don't need temporary strings here, instead, it'd be better to write them directly to `f`.